share/availability/discovery/discovery.go

```diff
@@ -94,36 +113,65 @@ func (d *Discovery) WithOnPeersUpdate(f OnUpdatedPeers) {
 	}
 }
 
+func (d *Discovery) triggerDiscovery() {
+	select {
+	case d.triggerDisq <- struct{}{}:
+	default:
+	}
+}
+
 // handlePeersFound receives peers and tries to establish a connection with them.
 // Peer will be added to PeerCache if connection succeeds.
-func (d *Discovery) handlePeerFound(ctx context.Context, topic string, peer peer.AddrInfo) {
-	if peer.ID == d.host.ID() || len(peer.Addrs) == 0 || d.set.Contains(peer.ID) {
-		return
-	}
-	err := d.set.TryAdd(peer.ID)
-	if err != nil {
-		log.Debug(err)
-		return
+func (d *Discovery) handlePeerFound(ctx context.Context, peer peer.AddrInfo) bool {
+	log := log.With("peer", peer.ID)
+	switch {
+	case peer.ID == d.host.ID():
+		log.Debug("skip handle: self discovery")
+		return false
+	case len(peer.Addrs) == 0:
+		log.Debug("skip handle: empty address list")
+		return false
+	case d.set.Size() >= d.set.Limit():
+		log.Debug("skip handle: enough peers found")
+		return false
+	case d.connector.HasBackoff(peer.ID):
+		log.Debug("skip handle: backoff")
+		return false
 	}
 
-	err = d.connector.Connect(ctx, peer)
-	if err != nil {
-		log.Debug(err)
-		d.set.Remove(peer.ID)
-		return
+	switch d.host.Network().Connectedness(peer.ID) {
+	case network.Connected:
+		d.connector.Backoff(peer.ID) // we still have to backoff the connected peer
+	case network.NotConnected:
+		err := d.connector.Connect(ctx, peer)
+		if err != nil {
+			log.Debugw("unable to connect", "err", err)
```

Comment:
log peer as well

Reply:
the logger is enhanced by peer field by that point.